### PR TITLE
docs(halyard/ha): Add Halyard doc page for High Availability.

### PR DESCRIPTION
**Pull request by joonlim**
image
image

### CODE REVIEW

**Comment by joonlim** [comment]
Please review @jtk54 @lwander @nhayes @dorbin

File: reference/halyard/high-availability.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
+---
+layout: single
+title:  "High Availability"
+sidebar:
+  nav: reference
+---
+
+{% include toc %}
+
+This page describes how you can configure a Halyard deployment to increase the availability of specific services beyond simply [horizontally scaling](/setup/productionize/scaling/horizontal-scaling/) the service. Halyard does this by splitting the functionalities of a service into separate logical services. The benefits of doing this is specific to the service that is being split.
```
Comment (jtk54) [inline reference/halyard/high-availability.md]: 'logical services' -> 'logical roles'? The service/services terminology is a bit confusing.

File: reference/halyard/high-availability.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+It is important to know that, when split, the new logical services are given new names. This means that these logical services can be configured and scaled independently of each other.
```
Comment (jtk54) [inline reference/halyard/high-availability.md]: `When split into different roles, the new logical...`

File: reference/halyard/high-availability.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+__Important:__ Please note that Halyard only supports this functionality for a [distributed Spinnaker deployment](/setup/install/environment/#distributed-installation) configured with a [manifest based Kubernetes account](/setup/install/providers/kubernetes-v2/).
+
+## HA Clouddriver
+
+Clouddriver benefits greatly from isolating its operations into separate services. To split Clouddriver for increased availability, run:
+
+```hal config deploy ha clouddriver enable```
+
+When Spinnaker is deployed with this flag enabled, Clouddriver will be deploy as three different services, each only performing a subset of the base Clouddriver's operations. Although by default the three Clouddriver services will communicate with the global Redis (all Spinnaker services speak to this Redis) provided by Halyard, it is recommended that the logical Clouddriver services be configured to communicate with an external Redis service. To be most effective, the Clouddriver-Read-Only should be configured to speak to a Redis slave, while the other two should be configured to speak to the master. This is handled by automatically by Halyard if the user provides the two endpoints using this command:
```
Comment (jtk54) [inline reference/halyard/high-availability.md]: s/deploy/deployed

s/slave/read replica/

File: reference/halyard/high-availability.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+The second logical Clouddriver service is the Clouddriver-Read-Write service, or ```clouddriver-rw```. This service handles all mutation operations aside from what the Clouddriver-Caching service does. This service can be scaled to handle an increased number of writes.
+
+This service's name when [configuring its sizing](/reference/halyard/component-sizing/) is ```spin-clouddriver-rw```.
+
+To add a [custom profile](/reference/halyard/custom/#custom-profiles) or [custom service settings](/reference/halyard/custom/#custom-service-settings) for this service, use the name ```clouddriver-rw```.
+
+### Clouddriver-Read-Only
+
+Inspired by [Netflix](https://blog.spinnaker.io/scaling-spinnaker-at-netflix-part-1-8a5ae51ee6de), the Clouddriver-Read-Only service, or ```clouddriver-ro``` handles all read requests to Clouddriver. This service can be scaled to handle an increased number of reads.
```
Comment (jtk54) [inline reference/halyard/high-availability.md]: I think each of these configurations come from Netflix's large scale experience described in that doc. I'd either mention it more holistically further up in the doc or I'd exclude it entirely.

File: reference/halyard/high-availability.md [hunk 5]
```diff
@@ -7,9 +7,9 @@ sidebar:
...
+This page describes how you can configure a Halyard deployment to increase the availability of specific services beyond simply [horizontally scaling](/setup/productionize/scaling/horizontal-scaling/) the service. Halyard does this by splitting the functionalities of a service into separate logical roles. The benefits of doing this is specific to the service that is being split. These deployment strategies are inspired by [Netflix's large scale experience](https://blog.spinnaker.io/scaling-spinnaker-at-netflix-part-1-8a5ae51ee6de).
 
-It is important to know that, when split, the new logical services are given new names. This means that these logical services can be configured and scaled independently of each other.
+When split into different roles, the new logical services are given new names. This means that these logical services can be configured and scaled independently of each other.
```
Comment (jtk54) [inline reference/halyard/high-availability.md]: Still not sold on the `services` terminology. `roles`? Thoughts on this?

**Review by jtk54 — APPROVED** [review]

File: reference/halyard/high-availability.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+title:  "High Availability"
+sidebar:
+  nav: reference
+---
+
+{% include toc %}
+
+This page describes how you can configure a Halyard deployment to increase the availability of specific services beyond simply [horizontally scaling](/setup/productionize/scaling/horizontal-scaling/) the service. Halyard does this by splitting the functionalities of a service into separate logical roles. The benefits of doing this is specific to the service that is being split. These deployment strategies are inspired by [Netflix's large scale experience](https://blog.spinnaker.io/scaling-spinnaker-at-netflix-part-1-8a5ae51ee6de).
+
+When split into different roles, the new logical services are given new names. This means that these logical services can be configured and scaled independently of each other.
```
Comment (lwander) [inline reference/halyard/high-availability.md]: I agree that "roles" is confusing since it's usually only associated with authz. I don't mind calling these separate 'services', since technically each specialization of clouddriver/echo is a new service with its own config & load balancer.

File: reference/halyard/high-availability.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+__Important:__ Please note that Halyard only supports this functionality for a [distributed Spinnaker deployment](/setup/install/environment/#distributed-installation) configured with a [manifest based Kubernetes account](/setup/install/providers/kubernetes-v2/).
+
+## HA Clouddriver
+
+Clouddriver benefits greatly from isolating its operations into separate services. To split Clouddriver for increased availability, run:
+
+```hal config deploy ha clouddriver enable```
+
+When Spinnaker is deployed with this flag enabled, Clouddriver will be deployed as three different services, each only performing a subset of the base Clouddriver's operations. Although by default the three Clouddriver services will communicate with the global Redis (all Spinnaker services speak to this Redis) provided by Halyard, it is recommended that the logical Clouddriver services be configured to communicate with an external Redis service. To be most effective, the Clouddriver-Read-Only should be configured to speak to a Redis read replica, while the other two should be configured to speak to the master. This is handled by automatically by Halyard if the user provides the two endpoints using this command:
```
Comment (lwander) [inline reference/halyard/high-availability.md]: I would link to redis documentation describing read replicas & such here.

File: reference/halyard/high-availability.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+The second logical Clouddriver service is the Clouddriver-Read-Write service, or ```clouddriver-rw```. This service handles all mutation operations aside from what the Clouddriver-Caching service does. This service can be scaled to handle an increased number of writes.
+
+This service's name when [configuring its sizing](/reference/halyard/component-sizing/) is ```spin-clouddriver-rw```.
+
+To add a [custom profile](/reference/halyard/custom/#custom-profiles) or [custom service settings](/reference/halyard/custom/#custom-service-settings) for this service, use the name ```clouddriver-rw```.
+
+### Clouddriver-Read-Only
+
+The Clouddriver-Read-Only service, or ```clouddriver-ro``` handles all read requests to Clouddriver. This service can be scaled to handle an increased number of reads.
```
Comment (lwander) [inline reference/halyard/high-availability.md]: For inline `code` use only a single tick (`)

File: reference/halyard/high-availability.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+The first of the three logical Clouddriver services is the Clouddriver-Caching service, or ```clouddriver-caching```. This service is responsible for the caching and retrieving of cloud infrastructure data. Since this is all that the Clouddriver-Caching is doing, there is no communication between this service and any other Spinnaker service.
+
+This service's name when [configuring its sizing](/reference/halyard/component-sizing/) is ```spin-clouddriver-caching```.
+
+To add a [custom profile](/reference/halyard/custom/#custom-profiles) or [custom service settings](/reference/halyard/custom/#custom-service-settings) for this service, use the name ```clouddriver-caching```.
+
+### Clouddriver-Read-Write
+
+The second logical Clouddriver service is the Clouddriver-Read-Write service, or ```clouddriver-rw```. This service handles all mutation operations aside from what the Clouddriver-Caching service does. This service can be scaled to handle an increased number of writes.
```
Comment (lwander) [inline reference/halyard/high-availability.md]: all mutation operations -> all mutating operations

Comment (lwander) [reply reference/halyard/high-availability.md]: I'd mention that this needs to talk to a write replica of redis as well, since it's also writing data into redis, not just the cloud provider.

**Review by lwander — APPROVED** [review]

**Comment by joonlim** [comment]
Added images for the split services. Is this helpful or unnecessary?
@jtk54 @lwander 

image

**Comment by joonlim** [comment]
@dorbin Would love if you can comment on the new color I chose for the architecture diagram.

File: reference/halyard/high-availability.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+  nav: reference
+---
+
+{% include toc %}
+
+This page describes how you can configure a Halyard deployment to increase the availability of specific services beyond simply [horizontally scaling](/setup/productionize/scaling/horizontal-scaling/) the service. Halyard does this by splitting the functionalities of a service into separate logical roles. The benefits of doing this is specific to the service that is being split. These deployment strategies are inspired by [Netflix's large scale experience](https://blog.spinnaker.io/scaling-spinnaker-at-netflix-part-1-8a5ae51ee6de).
+
+When split, the new logical services are given new names. This means that these logical services can be configured and scaled independently of each other.
+
+__Important:__ Please note that Halyard only supports this functionality for a [distributed Spinnaker deployment](/setup/install/environment/#distributed-installation) configured with a [manifest based Kubernetes account](/setup/install/providers/kubernetes-v2/).
```
Comment (dorbin) [inline reference/halyard/high-availability.md]: s/manifest based/manifest-based/

Is "account" the right word here? "provider"?

Comment (dorbin) [inline reference/halyard/high-availability.md]: Delete "Please note that "

File: reference/halyard/high-availability.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+ class clouddriver-caching,clouddriver-ro,clouddriver-rw,echo-scheduler,echo-slave split
+ </div>
+
+ {% include mermaid %}
+
+Clouddriver benefits greatly from isolating its operations into separate services. To split Clouddriver for increased availability, run:
+
+```hal config deploy ha clouddriver enable```
+
+When Spinnaker is deployed with this flag enabled, Clouddriver will be deployed as three different services, each only performing a subset of the base Clouddriver's operations. Although by default the three Clouddriver services will communicate with the global Redis (all Spinnaker services speak to this Redis) provided by Halyard, it is recommended that the logical Clouddriver services be configured to communicate with an external Redis service. To be most effective, the Clouddriver-Read-Only should be configured to speak to a Redis read replica, while the other two should be configured to speak to the master. This is handled by automatically by Halyard if the user provides the two endpoints using this command:
```
Comment (dorbin) [inline reference/halyard/high-availability.md]: After that first sentence in the paragrah, bullet-list the three logical services, linking them to their respective sections below, then the next sentence ("Although by default...") will be a separate paragraph.

File: reference/halyard/high-availability.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+```hal config deploy ha services clouddriver edit --redis-master-endpoint $REDIS_MASTER_ENDPOINT --redis-slave-endpoint $REDIS_SLAVE_ENDPOINT```
+
+More information on Redis replication can be [found here](https://redis.io/topics/replication).
+
+### Clouddriver-Caching
+
+The first of the three logical Clouddriver services is the Clouddriver-Caching service, or `clouddriver-caching`. This service is responsible for the caching and retrieving of cloud infrastructure data. Since this is all that the Clouddriver-Caching is doing, there is no communication between this service and any other Spinnaker service.
```
Comment (dorbin) [inline reference/halyard/high-availability.md]: "...of the three logical Clouddriver services is `clouddriver-caching`."

(Delete "the Clouddriver-Caching service, or ")

Comment (dorbin) [reply reference/halyard/high-availability.md]: In fact, you might as well make each of these three headings lower-case code-font.

### `clouddriver-caching`

Comment (dorbin) [reply reference/halyard/high-availability.md]: s/is responsible for the caching and retrieving of/caches and retrieves/

File: reference/halyard/high-availability.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+This service's name when [configuring its sizing](/reference/halyard/component-sizing/) is `spin-clouddriver-caching`.
```
Comment (dorbin) [inline reference/halyard/high-availability.md]: I'm confused by the difference between `clouddriver-caching` and `spin-clouddriver-caching`.

File: reference/halyard/high-availability.md [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+To add a [custom profile](/reference/halyard/custom/#custom-profiles) or [custom service settings](/reference/halyard/custom/#custom-service-settings) for this service, use the name `clouddriver-caching`.
+
+### Clouddriver-Read-Write
+
+The second logical Clouddriver service is the Clouddriver-Read-Write service, or `clouddriver-rw`. This service handles all mutating operations aside from what the Clouddriver-Caching service does. This service can be scaled to handle an increased number of writes.
```
Comment (dorbin) [inline reference/halyard/high-availability.md]: Similarly here, do we need a plain-English-hyphenated name and a code-font name? If we do, that's fine, and this sentence is fine, but it's simpler if we just call them `clouddriver-rw`, `clouddriver-caching`, `clouddriver-ro`.

File: reference/halyard/high-availability.md [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+ class clouddriver-caching,clouddriver-ro,clouddriver-rw,echo-scheduler,echo-slave split
+ </div>
+
+ {% include mermaid %}
+
+Echo can be split into two separate services that handle different operations. To split Echo for increased availability, run:
+
+```hal config deploy ha echo enable```
+
+When Spinnaker is deployed with this enabled, Echo will be deploy as two different services, one that handles Echo's scheduled cron-jobs and one that handles everything else. Although the logical Echo services still cannot be horizontally scaled, splitting the services will reduce the load on both.
```
Comment (dorbin) [inline reference/halyard/high-availability.md]: s/deploy/deployed/

File: reference/halyard/high-availability.md [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+layout: single
+title:  "High Availability"
+sidebar:
+  nav: reference
+---
+
+{% include toc %}
+
+This page describes how you can configure a Halyard deployment to increase the availability of specific services beyond simply [horizontally scaling](/setup/productionize/scaling/horizontal-scaling/) the service. Halyard does this by splitting the functionalities of a service into separate logical roles. The benefits of doing this is specific to the service that is being split. These deployment strategies are inspired by [Netflix's large scale experience](https://blog.spinnaker.io/scaling-spinnaker-at-netflix-part-1-8a5ae51ee6de).
+
```
Comment (dorbin) [inline reference/halyard/high-availability.md:11]: It looks like this is only for Clouddriver and Echo (currently?). Might be a good idea to mention that here at the top.

**Review by dorbin — CHANGES_REQUESTED** [review]
Good article and good feature, Joon!

File: reference/halyard/high-availability.md (hunk 10 above)
Comment (joonlim) [reply reference/halyard/high-availability.md]: changed to "provider"

File: reference/halyard/high-availability.md (hunk 13 above)
Comment (joonlim) [reply reference/halyard/high-availability.md]: `cloudddriver-caching` is the name used for a custom profile (e.g., `profiles/clouddriver-caching-local.yml`) and for service-settings overrides (e.g., `service-settings/clouddriver-caching.yml`).

`spin-clouddriver-caching` is the name used for to set the number of replicas and cpu/memory sizes.

Should I reword this to be more clear or are the links good enough?

File: reference/halyard/high-availability.md [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+Although by default the three Clouddriver services will communicate with the global Redis (all Spinnaker services speak to this Redis) provided by Halyard, it is recommended that the logical Clouddriver services be configured to communicate with an external Redis service. To be most effective, `clouddriver-ro` should be configured to speak to a Redis read replica, while the other two should be configured to speak to the master. This is handled automatically by Halyard if the user provides the two endpoints using this command:
+
+```hal config deploy ha services clouddriver edit --redis-master-endpoint $REDIS_MASTER_ENDPOINT --redis-slave-endpoint $REDIS_SLAVE_ENDPOINT```
+
+More information on Redis replication can be [found here](https://redis.io/topics/replication).
+
+### `clouddriver-caching`
+
+The first of the three logical Clouddriver services is the `clouddriver-caching` service. This service caches and retrieves of cloud infrastructure data. Since this is all that `clouddriver-caching` is doing, there is no communication between this service and any other Spinnaker service.
```
Comment (dorbin) [inline reference/halyard/high-availability.md]: Delete "of"

File: reference/halyard/high-availability.md (hunk 13 above)
Comment (dorbin) [reply reference/halyard/high-availability.md]: As discussed... :+1:

**Review by dorbin — APPROVED** [review]
:+1:
